src/usecase/feature_engineering.py

```diff
@@ -0,0 +1,294 @@
+    X_imputed = imputer.fit_transform(X)
+    
+    return X_imputed, y
+
+
+def outlier_detection(X: np.ndarray) -> pd.DataFrame:
+
+    """
+    This function detect outlier based on local density of points.
+    parameters
+    ----------
+    X: np.ndarray
+        X imputted will be use to detect outlier
+    
+    returns
+    -------
+    outlier_score: pd.Dataframe
+        every point of data has outlier score which is saved
+    """
+
+    LOF = LocalOutlierFactor()
+    Y_pred = LOF.fit_predict(X)
+    X_score = LOF.negative_outlier_factor_
+
+    outlier_score = pd.DataFrame()
+    outlier_score['score'] = X_score
+
+    return outlier_score
+
+
+def remove_outlier(
+    X_imputed_df: pd.DataFrame,
+    outlier_score: pd.DataFrame,
+    y: pd.DataFrame
+    ) -> Tuple[pd.DataFrame, pd.DataFrame]:
+
+    """
+    This function will remove outlier on the dataset
+    parameters
+    ----------
+    X_imputed_df: pd.DataFrame
+    outlier_score: pd.DataFrame
+    y: pd.DataFrame
+
+    returns
+    -------
+    X_result_outlier: pd.DataFrame
+        Dataset without outlier points (cleaned data)
+    y_result_outlier: pd.DataFrame
+        y vector cleaned
+    """
+
+    filt = outlier_score["score"] < -1.5
+    outlier_index = outlier_score[filt].index.tolist()
+    X_result_outlier = pd.DataFrame(X_imputed_df).drop(outlier_index)
+    y_result_outlier = y.drop(outlier_index).values
+
+    return X_result_outlier, y_result_outlier
+
+
+def pca_analysis(
+    dataframe: pd.DataFrame,
+    objective_variance: float
+    ) -> Tuple[any, np.ndarray]:
+
+    """
+    This function resolve the problem of multicolinearity in dataset. Some features are too
+    correlated. This can be redundant information that we need to remove.
+    parameters
+    ----------
+    dataframe: pd.DataFrame
+    objective_variance: float
+
+    returns
+    -------
+    pca_out: any
+        pca object
+    principalComponents: np.ndarray
+        array of principals components 
+    """
+    #normalisation step
+    dataframe = zscore(dataframe)
+
+    pca = PCA(n_components=objective_variance)
+    pca_out = pca.fit(dataframe)
+    principalComponents = pca.transform(dataframe)
+
+
+    return pca_out, principalComponents
+
+
+def create_final_dataset_with_pca(
+    pca_out: any, principalComponents: np.array,
+    y: pd.DataFrame
+    ) -> pd.DataFrame:
+
+    """
+    This function will create the final dataset by using principals components and adding label.
+    parameters
+    ----------
+    pca_out: any
+        pca object
+    principalComponents: np.ndarray
+        array of principals components 
+    y: pd.Dataframe
+        y label vector
+
+    returns
+    -------
+    pca_df: pd.DataFrame
+        dataframe with principals components and label
+    """
+
+    #construction of dataset
+    new_size_dataset = pca_out.components_.shape[1]
+    list_colmuns = []
+    for i in range(1, new_size_dataset):
+        list_colmuns.append("PC"+str(i))
```

Comment:
Tu peux faire une liste compréhension à la place (plus rapide qu'une boucle for) : 
["PC"+str(i) for i in range(1,nb_of_variables)]

un peu de lecture : https://python.doctor/page-comprehension-list-listes-python-cours-debutants ;)